exec/java-exec/src/main/java/org/apache/drill/exec/store/avro/AvroSchemaUtil.java

```diff
@@ -0,0 +1,274 @@
+        .addContext("Column", columnName)
+        .build(logger);
+    }
+    List<Schema> unionSchemas = schema.getTypes();
+
+    // exclude all schemas with null type
+    List<Schema> nonNullSchemas = unionSchemas.stream()
+      .filter(unionSchema -> !Schema.Type.NULL.equals(unionSchema.getType()))
+      .collect(Collectors.toList());
+
+    // if original schema has two elements and only one non-nullable schema, this is simple nullable type
+    if (unionSchemas.size() == 2 && nonNullSchemas.size() == 1) {
+      return nonNullSchemas.get(0);
+    } else {
+      return throwUnsupportedErrorForType("complex union", columnName);
+    }
+  }
+
+  private static <T> T throwUnsupportedErrorForType(String type, String columnName) {
+    throw UserException.unsupportedError()
+      .message("'%s' type is not supported", type)
+      .addContext("Column", columnName)
+      .build(logger);
+  }
+
+  /**
+   * Class is responsible for converting Avro schema into Drill metadata description of the schema.
+   * It does not hold state and thus is thread-safe.
+   */
+  private static class SchemaConverter {
+
+    private static final SchemaConverter INSTANCE = new SchemaConverter();
+
+    TupleMetadata convert(Schema schema) {
+      /*
+        Avro allows to reference types by name, sometimes reference can be done to the type under construction.
+        For example, a linked-list of 64-bit values:
+        {
+          "type": "record",
+          "name": "LongList",
+          "fields" : [
+             {"name": "value", "type": "long"},             // each element has a long
+             {"name": "next", "type": ["null", "LongList"]} // optional next element
+           ]
+        }
+
+        Since we cannot build record type which is not constructed yet, when such situation is detected,
+        record type is set to Drill Map without columns, columns will be detected when reading actual data.
+
+        `typeNamesUnderConstruction` is a holder to store record type names under construction to detect
+         reference to the types which are not yet constructed.
+       */
+      Set<String> typeNamesUnderConstruction = new HashSet<>();
+      TupleSchema tupleSchema = new TupleSchema();
+
+      // add current record type to the set of types under construction
+      typeNamesUnderConstruction.add(schema.getFullName());
+
+      List<Schema.Field> fields = schema.getFields();
+      fields.stream()
+        .map(field -> convert(field, typeNamesUnderConstruction))
+        .forEach(tupleSchema::add);
+      return tupleSchema;
+    }
+
+    private ColumnMetadata convert(Schema.Field field, Set<String> typeNamesUnderConstruction) {
+      Schema fieldSchema = field.schema();
+      return defineColumn(field.name(), fieldSchema, TypeProtos.DataMode.REQUIRED, typeNamesUnderConstruction);
+    }
+
+    private ColumnMetadata defineColumn(String name, Schema fieldSchema,
```

Comment:
Are there any cases where two distinct Avro types or variations map to a single Drill type? If so, you'd later want to create distinct `ColumnConverter`s for each variation. This might be a handy place to use `ColumnMetadata` properties to capture those nuances to be used later when creating the column converter.